include/shell/shell.h

```diff
@@ -1034,6 +1035,30 @@ int shell_execute_cmd(const struct shell *shell, const char *cmd);
  */
 int shell_set_root_cmd(const char *cmd);
 
+/**
+ * @brief Allow application to control whether terminal output uses colored
+ * syntax.
+ *
+ * @param[in] shell	Pointer to the shell instance.
+ * @param[in] use_colors	Color mode.
+ *
+ * @retval 0 if set.
+ * @retval -EINVAL if shell is NULL.
+ */
+int shell_use_colors_set(const struct shell *shell, bool use_colors);
+
+/**
+ * @brief Allow application to control whether user input is obscured with
+ * asterisks -- useful for implementing passwords.
+ *
+ * @param[in] shell	Pointer to the shell instance.
+ * @param[in] obscure	Obscure mode.
+ *
+ * @retval 0 if set.
+ * @retval -EINVAL if shell is NULL.
+ */
+int shell_obscure_set(const struct shell *shell, bool obscure);
```

Comment:
I'd somewhat prefer if these APIs handled nesting properly somehow.

As-is, if you have a function that tries to temporarily have obscured input or disabled colors that's called from a context that already has obscured input or disabled colors you're going to have a bad time.

Reply:
Well, how about either also adding getters, so the application can restore the previous setting?

Reply:
> Well, how about also adding getters, so the application can restore the previous setting?

It'd ideally be a setter+return previous, like how `arch_irq_lock` does it.

Otherwise there's a race condition if another thread preempts you between you getting the old value and setting a new one.

Reply:
I've added a new commit with changes to enable atomic set of flags and return the previous value.  I went ahead and added a few more public setters using this.